include/ttmlir/Dialect/TTNN/IR/TTNNWorkarounds.h

```diff
@@ -0,0 +1,154 @@
+    } else if constexpr (std::is_same<T, BufferType>::value) {
+      tensorBufferTypeWorkaround = std::make_optional(workaround);
+    } else if constexpr (std::is_same<T, TensorMemoryLayout>::value) {
+      tensorMemoryLayoutWorkaround = std::make_optional(workaround);
+    }
+    return *this;
+  }
+
+  // Equality operator.
+  bool operator==(const TTNNOperandWorkarounds &rhs) const {
+    return tensorLayoutWorkaround == rhs.tensorLayoutWorkaround &&
+           tensorBufferTypeWorkaround == rhs.tensorBufferTypeWorkaround &&
+           tensorMemoryLayoutWorkaround == rhs.tensorMemoryLayoutWorkaround;
+  }
+
+  // Inequality operator.
+  bool operator!=(const TTNNOperandWorkarounds &rhs) const {
+    return !(*this == rhs);
+  }
+
+  // Returns true if any of the workarounds is set.
+  bool hasAnyWorkaround() {
+    return tensorLayoutWorkaround || tensorBufferTypeWorkaround ||
+           tensorMemoryLayoutWorkaround;
+  }
+
+private:
+  // Tensor layout workaround
+  TensorLayoutWorkaround tensorLayoutWorkaround;
+  // Tensor buffer type workaround
+  TensorBufferTypeWorkaround tensorBufferTypeWorkaround;
+  // Tensor memory layout workaround.
+  TensorMemoryLayoutWorkaround tensorMemoryLayoutWorkaround;
+};
+
+// Class that encapsulates operands workarounds.
+// It contains input and output workarounds for operands.
+class TTNNOperandsWorkarounds {
+public:
+  // Default constructor with no workarounds.
+  TTNNOperandsWorkarounds() {}
+
+  // Constructor that takes input and output workarounds for operands.
+  TTNNOperandsWorkarounds(
+      std::vector<TTNNOperandWorkarounds> inputOperandWorkarounds,
+      std::vector<TTNNOperandWorkarounds> outputOperandWorkarounds)
+      : inputOperandWorkarounds(inputOperandWorkarounds),
+        outputOperandWorkarounds(outputOperandWorkarounds) {}
+
+  // Returns input operand workarounds.
+  std::vector<TTNNOperandWorkarounds> getInputOperandWorkarounds() const {
+    return inputOperandWorkarounds;
+  }
+
+  // Returns output operand workarounds.
+  std::vector<TTNNOperandWorkarounds> getOutputOperandWorkarounds() const {
+    return outputOperandWorkarounds;
+  }
+
+  // Adds input operand workaround.
+  TTNNOperandsWorkarounds &
+  addInputOperandWorkaround(TTNNOperandWorkarounds inputOperandWorkaround) {
+    inputOperandWorkarounds.emplace_back(inputOperandWorkaround);
+    return *this;
+  }
+
+  // Adds output operand workaround.
+  TTNNOperandsWorkarounds &
+  addOutputOperandWorkaround(TTNNOperandWorkarounds outputOperandWorkaround) {
+    outputOperandWorkarounds.emplace_back(outputOperandWorkaround);
+    return *this;
+  }
+
+private:
+  // Workarounds for input operands.
+  std::vector<TTNNOperandWorkarounds> inputOperandWorkarounds;
+  // Workarounds for output operands.
+  std::vector<TTNNOperandWorkarounds> outputOperandWorkarounds;
+};
+
+// Class that provides factory methods for creating workarounds.
+class WorkaroundFactory {
```

Comment:
Why do we need a class? Why not just function have functions?

Reply:
At first it was a bit of mess and complex hirearchy of classes, hence I opted out for this. I agree, this is not needed anymore, thanks for pointing out!